### PR TITLE
GetBlock() - Simple optimisations

### PR DESCRIPTION
**Pull request by jmozah**
**What type of PR is this?**
 Optimisation

**What does this PR do? Why is it needed?**
During debugging of https://github.com/prysmaticlabs/prysm/issues/8943, it was found that some of the operations done
during GetBlock() can be optimised further. This PR does the simple optimisations under a flag.

**Which issues(s) does this PR fix?**
Related to  #8943 

**Other notes for review**
The flag is --enable-get-block-optimizations, and it is disabled by default.

### CODE REVIEW

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 1]
```diff
@@ -142,6 +137,70 @@ func (vs *Server) GetBlock(ctx context.Context, req *ethpb.BlockRequest) (*ethpb
...
+}
+
+func (vs *Server) optimizedDepositAndPackAttestations(ctx context.Context, head state.BeaconState, eth1Data *eth.Eth1Data) ([]*eth.Deposit, []*eth.Attestation, error) {
+	eg, egctx := errgroup.WithContext(ctx)
+	var deposits []*eth.Deposit
+	var atts []*eth.Attestation
+
+	eg.Go(func() error {
+		// Pack ETH1 deposits which have not been included in the beacon chain.
+		localDeposits, err := vs.deposits(ctx, head, eth1Data)
```
Comment (prestonvanloon) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:171]: Shouldn't you use egctx here?
```suggestion
		localDeposits, err := vs.deposits(egctx, head, eth1Data)
```

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 2]
```diff
@@ -142,6 +137,70 @@ func (vs *Server) GetBlock(ctx context.Context, req *ethpb.BlockRequest) (*ethpb
...
+			return egctx.Err()
+		default:
+		}
+		deposits = localDeposits
+		return nil
+	})
+
+	eg.Go(func() error {
+		// Pack aggregated attestations which have not been included in the beacon chain.
+		localAtts, err := vs.packAttestations(ctx, head)
```
Comment (prestonvanloon) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:187]: Here too
```suggestion
		localAtts, err := vs.packAttestations(egctx, head)
```

File: shared/featureconfig/flags.go [hunk 3]
```diff
@@ -135,6 +135,11 @@ var (
...
 		Usage: "This enables active balance cache cache to improve node performance during block processing",
 	}
+
+	enableGetBlockOptimizations = &cli.BoolFlag{
+		Name:  "enable-get-block-optimizations",
+		Usage: "This enables some optimizations on the GetBlock() function.",
+	}
 )
 
 // devModeFlags holds list of flags that are set when development mode is on.
```
Comment (prestonvanloon) [inline shared/featureconfig/flags.go:145]: Please add your flag to devModeFlags, if it's safe to test in testnets

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 4]
```diff
@@ -209,15 +268,17 @@ func (vs *Server) eth1DataMajorityVote(ctx context.Context, beaconState state.Be
...
-	lastBlockByEarliestValidTime, err := vs.Eth1BlockFetcher.BlockByTimestamp(ctx, earliestValidTime)
-	if err != nil {
-		log.WithError(err).Error("Could not get last block by earliest valid time")
-		return vs.randomETH1DataVote(ctx)
-	}
-	// Increment the earliest block if the original block's time is before valid time.
-	// This is very likely to happen because BlockTimeByHeight returns the last block AT OR BEFORE the specified time.
-	if lastBlockByEarliestValidTime.Time < earliestValidTime {
-		lastBlockByEarliestValidTime.Number = big.NewInt(0).Add(lastBlockByEarliestValidTime.Number, big.NewInt(1))
+	if !featureconfig.Get().EnableGetBlockOptimizations {
```
Comment (prestonvanloon) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:271]: Why is this branch omitted when the flag is on?

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 5]
```diff
@@ -21,6 +21,7 @@ import (
 	"github.com/prysmaticlabs/prysm/beacon-chain/core/state/interop"
 	"github.com/prysmaticlabs/prysm/beacon-chain/state"
 	dbpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+	eth "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
 	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
```
Comment (prestonvanloon) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:25]: These are the same imports, please use only one

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go (hunk 4 above)
Comment (jmozah) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:271]: This branch seems to not do anything here.  
`lastBlockByEarliestValidTime` is never used anywhere inside the code.
I need a confirmation if my thinking is right.
WDYT @nisdas

Comment (nisdas) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:271]: That is a bit more complicated, as that value was used for our `inRangeVotes` method. This was removed previously when we had the voting bug and defaulted to independent voting compared to voting by majority. Do we ever expect to add the vote by majority feature back ? @prestonvanloon

Comment (jmozah) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:486]: My 2 cents: Even if want to add the feature back sometime later. Let's remove unused code for now.

**Review by terencechain — DISMISSED** [review]
Can we implement benchmarks on before and after? We should know how much this improves

Numbers from tracing instrumentation will be helpful too

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go (hunk 4 above)
Comment (nisdas) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:486]: Discussed offline, its better to just remove it now. Performing the extra BlockByTimeStamp call is expensive and can increase block proposal times needlessly.

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 6]
```diff
@@ -230,16 +232,9 @@ func (vs *Server) buildPhase0BlockData(ctx context.Context, req *ethpb.BlockRequ
 		return nil, fmt.Errorf("could not get ETH1 data: %v", err)
 	}
 
-	// Pack ETH1 deposits which have not been included in the beacon chain.
-	deposits, err := vs.deposits(ctx, head, eth1Data)
+	deposits, atts, err := vs.depositAndPackAttestations(ctx, head, eth1Data)
```
Comment (nisdas) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:235]: it is a bit odd to group deposit and attestation packing into one method. But I do understand that it makes it cleaner when applying the flag.

**Comment by jmozah** [comment]
> Can we implement benchmarks on before and after? We should know how much this improves
> 
> Numbers from tracing instrumentation will be helpful too

@terencechain yeah we can do that.

**Comment by jmozah** [comment]
> > Can we implement benchmarks on before and after? We should know how much this improves
> > Numbers from tracing instrumentation will be helpful too
> 
> @terencechain yeah we can do that.

@terencechain On the second thought, i think it will be nice to run the instrumentation and monitor block proposal times, after we fix the rest of the fixes related to getBlock(). What say?

**Comment by terencechain** [comment]
> > > Can we implement benchmarks on before and after? We should know how much this improves
> > > Numbers from tracing instrumentation will be helpful too
> > 
> > 
> > @terencechain yeah we can do that.
> 
> @terencechain On the second thought, i think it will be nice to run the instrumentation and monitor block proposal times, after we fix the rest of the fixes related to getBlock(). What say?

That's fine with me. Let's wait to merge this until we can monitor block proposal time then

Another method here (easier) is to run getblock benchmark test on both featureflag on/off

**Comment by jmozah** [comment]
> Another method here (easier) is to run getblock benchmark test on both featureflag on/off

Thats a nice idea. I will do this.

**Comment by jmozah** [comment]
> > Another method here (easier) is to run getblock benchmark test on both featureflag on/off
> 
> Thats a nice idea. I will do this.

Decided offline with @terencechain to keep this PR open until we fix other issues in GetBlock()

**Comment by jmozah** [comment]
This PR will be closed once all the tracking issues are closed.

**Comment by jmozah** [comment]
@terencechain I think it was a bad idea to hold this PR until i fix other things.  This is very painful and excruciating to resolve merge conflicts and maintain it. So it did the benchmarking's like you recommended and want to close this PR.

 Lets track other problems in separate Issue and fix them one by one as we get time.

**Comment by jmozah** [comment]
@terence 
I think it was a bad idea to hold this PR until I fix other things. This is very painful and excruciating to resolve merge conflicts and maintain them. So I did the benchmarking's like you recommended [here](https://github.com/prysmaticlabs/prysm/pull/9541#issuecomment-918320748) and want to close this PR if you approve.

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 7]
```diff
@@ -423,15 +480,17 @@ func (vs *Server) eth1DataMajorityVote(ctx context.Context, beaconState state.Be
...
-		lastBlockByEarliestValidTime.Number = big.NewInt(0).Add(lastBlockByEarliestValidTime.Number, big.NewInt(1))
+	if !features.Get().EnableGetBlockOptimizations {
+		lastBlockByEarliestValidTime, err := vs.Eth1BlockFetcher.BlockByTimestamp(ctx, earliestValidTime)
+		if err != nil {
+			log.WithError(err).Error("Could not get last block by earliest valid time")
+			return vs.randomETH1DataVote(ctx)
+		}
+		// Increment the earliest block if the original block's time is before valid time.
+		// This is very likely to happen because BlockTimeByHeight returns the last block AT OR BEFORE the specified time.
+		if lastBlockByEarliestValidTime.Time < earliestValidTime {
```
Comment (nisdas) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:491]: if this value isnt being used anywhere you can remove it, no need to gate it behind the flag.

Comment (jmozah) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:491]: makes sense

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 8]
```diff
@@ -278,6 +271,70 @@ func (vs *Server) ProposeBeaconBlock(ctx context.Context, req *ethpb.GenericSign
 	return vs.proposeGenericBeaconBlock(ctx, blk)
 }
 
+func (vs *Server) depositAndPackAttestations(ctx context.Context, head state.BeaconState, eth1Data *ethpb.Eth1Data) ([]*ethpb.Deposit, []*ethpb.Attestation, error) {
```
Comment (terencechain) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:274]:
```suggestion
func (vs *Server) packDepositsAndAttestations(ctx context.Context, head state.BeaconState, eth1Data *ethpb.Eth1Data) ([]*ethpb.Deposit, []*ethpb.Attestation, error) {
```

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 9]
```diff
@@ -278,6 +271,70 @@ func (vs *Server) ProposeBeaconBlock(ctx context.Context, req *ethpb.GenericSign
...
+	// Pack aggregated attestations which have not been included in the beacon chain.
+	atts, err := vs.packAttestations(ctx, head)
+	if err != nil {
+		return nil, nil, status.Errorf(codes.Internal, "Could not get attestations to pack into block: %v", err)
+	}
+
+	return deposits, atts, nil
+}
+
+func (vs *Server) optimizedDepositAndPackAttestations(ctx context.Context, head state.BeaconState, eth1Data *ethpb.Eth1Data) ([]*ethpb.Deposit, []*ethpb.Attestation, error) {
```
Comment (terencechain) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:298]:
```suggestion
func (vs *Server) optimizedPackDepositsAndAttestations(ctx context.Context, head state.BeaconState, eth1Data *ethpb.Eth1Data) ([]*ethpb.Deposit, []*ethpb.Attestation, error) {
```

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 10]
```diff
@@ -278,6 +271,70 @@ func (vs *Server) ProposeBeaconBlock(ctx context.Context, req *ethpb.GenericSign
...
+	var deposits []*ethpb.Deposit
+	var atts []*ethpb.Attestation
+
+	eg.Go(func() error {
+		// Pack ETH1 deposits which have not been included in the beacon chain.
+		localDeposits, err := vs.deposits(egctx, head, eth1Data)
+		if err != nil {
+			return status.Errorf(codes.Internal, "Could not get ETH1 deposits: %v", err)
+		}
+		// if the original context is cancelled, then cancel this routine toobazel run //:gazelle -- fix`
```
Comment (terencechain) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:309]:
```suggestion
		// if the original context is cancelled, then cancel this routine too
```

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 11]
```diff
@@ -551,7 +605,7 @@ func (vs *Server) deposits(
 	if vs.MockEth1Votes || !vs.Eth1InfoFetcher.IsConnectedToETH1() {
 		return []*ethpb.Deposit{}, nil
 	}
-	// Need to fetch if the deposits up to the state's latest eth 1 data matches
+	// Need to fetch if the deposits up to the state's latest ethpb 1 data matches
```
Comment (terencechain) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:608]:
```suggestion
	// Need to fetch if the deposits up to the state's latest eth 1 data matches
```

File: config/features/config.go [hunk 12]
```diff
@@ -50,6 +50,7 @@ type Flags struct {
 	EnableOptimizedBalanceUpdate        bool // EnableOptimizedBalanceUpdate uses an updated method of performing balance updates.
 	EnableDoppelGanger                  bool // EnableDoppelGanger enables doppelganger protection on startup for the validator.
 	EnableHistoricalSpaceRepresentation bool // EnableHistoricalSpaceRepresentation enables the saving of registry validators in separate buckets to save space
+	EnableGetBlockOptimizations         bool // EnableGetBlockOptimizations optimizes some elements of the GetBlock() function
```
Comment (terencechain) [inline config/features/config.go:53]:
```suggestion
	EnableGetBlockOptimizations         bool // EnableGetBlockOptimizations optimizes some elements of the GetBlock() function.
```

**Review by nisdas — APPROVED** [review]

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 13]
```diff
@@ -551,7 +605,7 @@ func (vs *Server) deposits(
 	if vs.MockEth1Votes || !vs.Eth1InfoFetcher.IsConnectedToETH1() {
 		return []*ethpb.Deposit{}, nil
 	}
-	// Need to fetch if the deposits up to the state's latest eth 1 data matches
+	// Need to fetch if the deposits up to the state's latest ethEnableGetBlockOptimizations 1 data matches
```
Comment (nisdas) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:608]:
```suggestion
	// Need to fetch if the deposits up to the state's latest eth1 data matches
```

**Review by nisdas — APPROVED** [review]

**Pull request merged**